src/main/scala/com/advancedtelematic/campaigner/db/Repository.scala

```diff
@@ -92,23 +92,33 @@ protected [db] class DeviceUpdateRepository()(implicit db: Database, ec: Executi
       }
     }
 
-  protected [db] def setUpdateStatusAction(update: UpdateId, device: DeviceId, status: DeviceStatus): DBIO[Unit] =
+  def findFailedByFailureCode(campaignId: CampaignId, failureCode: String): Future[Set[DeviceId]] = db.run {
+    Schema.deviceUpdates
+      .filter(_.campaignId === campaignId)
+      .filter(_.status === DeviceStatus.failed)
+      .filter(_.failureCode.map(_ === failureCode))
+      .map(_.deviceId)
+      .result
+      .map(_.toSet)
+  }
+
+  protected [db] def setUpdateStatusAction(update: UpdateId, device: DeviceId, status: DeviceStatus, failureCode: Option[String]): DBIO[Unit] =
     Schema.deviceUpdates
       .filter(_.updateId === update)
       .filter(_.deviceId === device)
-      .map(_.status)
-      .update(status)
+      .map(du => du.status -> du.failureCode)
```

Comment:
I'd personally prefer to use plain tuple `(du.status, du.failureCode)` instead of an arrow operator here as it looks more readable to me. According to http://slick.lightbend.com/doc/3.2.0/sql-to-slick.html#update, it seems to look more readable to Slick authors as well. It is a matter of taste, however, and since I'm not aware of any code style guidelines used in our projects, you are free to dismiss my comment.

Reply:
Indeed, that seems more readable here. I will update it.

Reply:
Done.